mac80211: ath9k: enable adding wireless noise to kernel entropy pool

alanswanson opened the pull request:
Fixes FS#1444
Signed-off-by: Alan Swanson <reiver@improbability.net>

hauke commented on package/kernel/mac80211/ath.mk:
```diff
@@ -206,6 +208,11 @@ endef
 
 define KernelPackage/ath9k/config
 
+	config ATH9K_HWRNG
```
This also depends on HW_RANDOM which is added by kmod-random-core.

hauke commented on package/kernel/mac80211/ath.mk:
```diff
@@ -206,6 +208,11 @@ endef
+	config ATH9K_HWRNG
+		bool "Add wireless noise as source of randomness to kernel entropy pool"
+		depends on PACKAGE_kmod-ath9k
+		default y
```
Please do not make this default y for now

CodeFetch reviewed (approved):
Nice! Should be tested for a while and might be a cool default feature.
Even if the additional entropy would be of bad quality, it does not impose security risks on Linux as entropy just feeds a prng. The code looks good. Thus I think this can be merged as is.

adrianschmutzler commented:
Hi, please use "mac80211:" prefix for your commit title, maybe similar to the PR title if that's not too long.
Despite, please add a description of what you do and why you do it to your commit message!

alanswanson commented:
Prefix added. The commit doesn't need further description as the title explains it. Otherwise I just be writing "Everybody's dead, Dave" multiple times differently.

adrianschmutzler commented:
The title maybe tells what you do, but definitely not why this is needed.

mx2345 commented:
Have been using this config option with my UAP-AC-LR (ath79) since September and it has worked just fine.

CodeFetch commented:
@adrianschmutzler Can you merge it now?

adrianschmutzler commented:
@CodeFetch I've removed the "needs changes" label I added before, but I won't decide on the content (I don't do kernel-related review).

alanswanson committed:
mac80211: ath9k: enable adding wireless noise to kernel entropy pool

This option allows adding the ath9k ADC register output as a source
of randomness into the Linux entropy pool at sufficient quality
random data (at least 10 bits and up to 22 bits of min-entropy for
a 32-bit value).

Fixes FS#1444
Signed-off-by: Alan Swanson <reiver@improbability.net>

alanswanson commented:
Just a fix for package select and depend.

Note there has been another user Rui Salvaterra submitted the same on [forum](https://forum.openwrt.org/t/feeding-the-entropy-pool-with-adc-noise-ath9k/53527) and mailing list. Felix noted [noted](https://lists.infradead.org/pipermail/openwrt-devel/2020-February/021666.html) issues with stuck beacons, hardware resets, connection stability issues. However testing has reveal no issues.

The file package/kernel/mac80211/patches/ath/543-ath9k_entropy_from_adc.patch could be culled as it only provides once off entropy versus newer urngd and optionally this.

CodeFetch commented:
@alanswanson Please add the HW_RANDOM dependency.
Citing from Felix:

> I think the benefit of continuous entropy collection is simply not worth
> the extra cost of potential stability issues and debugging time.
> (...)
> The issues that were reported years ago were mainly stuck beacons,
> increase in hardware resets, connection stability issues.

Indeed stressing the ath9k chip too much is not what we want, but more entropy for these devices would be great...

There was an idea which would is very interesting for embedded devices:

> Instead, I had a compile-time option to choose a number 0-32
> for how much entropy to assume a 32-bit value from the HWRNG
> contains. Default was something less than 32. I debated values
> in the 24-30 range, don't recall what I chose & don't think it
> Matters hugely.
https://lwn.net/Articles/760614/

But merging the PR would be okay I guess, as it just adds the option (maybe briefly state that it can lead to timing problems in the description).

alanswanson commented:
The dependency on HW_RANDOM is already set by selecting kmod-random-core per the first comment/review by hauke (which I'd missed marking conversation as resolved, now done).

I cannot find any timing issues raised in previous mailing lists discussions. Please provide a reference.

CodeFetch commented:
> I cannot find any timing issues raised in previous mailing lists discussions. Please provide a reference.

Stuck beacons ordinarily happen when the timing period of the beacon queue is being missed. If there are too many misses the chip is being reseted.

adrianschmutzler commented:
This has been merged in a4c0767fbc50e.
